src/Http/Routing/src/Builder/EndpointRoutingApplicationBuilderExtensions.cs

```diff
@@ -96,10 +106,22 @@ public static IApplicationBuilder UseEndpoints(this IApplicationBuilder builder,
             //
             // Each middleware gets its own collection of data sources, and all of those data sources also
             // get added to a global collection.
-            var routeOptions = builder.ApplicationServices.GetRequiredService<IOptions<RouteOptions>>();
-            foreach (var dataSource in endpointRouteBuilder.DataSources)
+            // In the global endpoint route case we only want to copy data sources once, so we wait for a specific key before copying data sources
+            // like in the case of minimal hosting
+            if (builder.Properties.TryGetValue(GlobalEndpointBuilderCopyRoutesKey, out _) ||
+                !builder.Properties.TryGetValue(GlobalEndpointRouteBuilderKey, out _))
+            {
+                var routeOptions = builder.ApplicationServices.GetRequiredService<IOptions<RouteOptions>>();
+                foreach (var dataSource in endpointRouteBuilder.DataSources)
+                {
+                    routeOptions.Value.EndpointDataSources.Add(dataSource);
+                }
+            }
+
+            // REVIEW: this 'if' could be removed, see comment in WebApplicationBuilder
+            if (!builder.Properties.TryGetValue(GlobalEndpointRouteBuilderKey, out _))
             {
-                routeOptions.Value.EndpointDataSources.Add(dataSource);
+                builder.Properties.Remove(EndpointRouteBuilder);
```

Comment:
Why does this need to removed in any case? I thought the idea was that when `__GlobalEndpointRouteBuilder` is set, we don't touch `__EndpointRouteBuilder` which might be used by a startup filter or something.

Reply:
We don't want branches to be affected by global routing.

Reply:
In my current local changes I moved the property removing logic into WebApplicationBuilder

Reply:
Makes sense. I remember the other thing we discussed to fix the branching was not including the `__GlobalEndpointRouteBuilder` key when enumerating properties, but moving the logic into `WebApplicationBuilder` sounds a lot cleaner.